### PR TITLE
style: Cleanup a bit the cascade.

### PR DESCRIPTION
**Pull request by emilio**
This makes an element available in StyleAdjuster, and uses it to replace some of our CascadeFlags (the ones that don't represent restrictions on what's actually cascaded, actually).

That fixes handwaviness in the handling of those flags from style reparenting,
and code duplication to handle tricky stuff like :visited.

There are a number of other changes that are worth noticing:

 * skip_root_and_item_based_display_fixup is renamed to skip_item_display_fixup:

   TElement::is_root() already implies being the document element, which by
   definition is not native anonymous and not a pseudo-element.

   Thus, you never get fixed-up if your NAC or a pseudo, which is what the code
   tried to avoid, so the only fixup with a point is the item one, which is
   necessary.

 * The pseudo-element probing code was refactored to return early a
   Option::<CascadeInputs>::None, which is nicer than what it was doing.

 * The visited_links_enabled check has moved to selector-matching time. The rest
   of the checks aren't based on whether the element is a link, or are properly
   guarded by parent_style.visited_style().is_some() or visited_rules.is_some().

   Thus you can transitively infer that no element will end up with a :visited
   style, not even from style reparenting.

Anyway, the underlying reason why I want the element in StyleAdjuster is because
we're going to implement an adjustment in there depending on the tag of the
element (converting display: contents to display: none depending on the tag), so
computing that information eagerly, including a hash lookup, wouldn't be nice.

See each commit for details.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19842)

### CODE REVIEW

**Comment by emilio** [comment]
https://treeherder.mozilla.org/#/jobs?repo=try&revision=7194d0303f872fd5aab32defa1909112ac75c2a6 is mostly-green (the assertion that those tests hit is the one from the canvas code that I had to uncomment in the end).

File: components/style/style_resolver.rs [hunk 1]
```diff
@@ -291,29 +291,38 @@ where
...
+            inputs,
+            &self.context.shared.guards,
+            parent_style,
+            parent_style,
+            layout_parent_style,
+            &self.context.thread_local.font_metrics_provider,
+            Some(&self.context.thread_local.rule_cache),
+            &mut conditions,
+        );
+
```
Comment (jryans) [inline components/style/style_resolver.rs]: Nit: seems like only one blank line is needed?

**Review by jryans — APPROVED** [review]
Thanks! I reviewed the last commit (about visited) and it looks good overall to me.

**Pull request merged**